### ISSUE TITLE
Instagram deprecating older oEmbed endpoint

### ISSUE THREAD

**Issue by chris-hinds** [ruby-oembed/ruby-oembed#74]
Instagram are deprecating their now "legacy" oEmbed endpoint on October 24th.

https://developers.facebook.com/docs/instagram/oembed-legacy

Is this library setup to use the newer api endpoint? If not will it be updated soon?

Looks like the Facebook oEmbed endpoint is being deprecated too

**Comment by metavida** [comment]
Ah! Thanks for the heads up! I haven't been following oEmbed news so missed this deprecation warning.

> Is this library setup to use the newer api endpoint? If not will it be updated soon?

As of v0.13.1 the official `OEmbed::Providers::Instagram` class [is **using the old endpoint**](https://github.com/ruby-oembed/ruby-oembed/blob/v0.13.1/lib/oembed/providers.rb#L206).

Reading through [the documentation for the new oEmbed endpoint](https://developers.facebook.com/docs/instagram/oembed) it seems that each request will now need to include an access token in the query param. I do not want to create a Facebook developer account nor would I be willing to distribute my app's access token along with this open-source library, so I don't see a way to support the new Instagram provider out of the box.

That said, this library definitely _can_ support you creating your own Instagram provider that includes your personal access token. It'd look something like this:

```ruby
# Create your instagram provider with its access_token set via environment variable
my_instagram = OEmbed::Provider.new("https://graph.facebook.com/v8.0/instagram_oembed?access_token=#{ENV['MY_INSTA_TOKEN']}")
my_instagram << "https://www.instagram.com/p/*"
my_instagram << "https://www.instagram.com/tv/*"

# Then, optionally, register your provider with the library
OEmbed::Providers.register(my_instagram)
```

I *would* consider a new API that would somehow allow people to provider their own access token. I just don't quite know how I would want to pull this off. Perhaps the built-in provider would be registered only if specially named environment variable like `ENV['OEMBED_INSTAGRAM_TOKEN']` is set? Perhaps the token could be passed in when calling `register_all` like `OEmbed::Providers.register_all(access_tokens: {instagram: ENV['MY_INSTA_TOKEN']})` or something??

I'd be interested to hear your feedback on what might make the most sense to you.

**Comment by jhass** [comment]
I think some providers, those that require configuration, not being passable to `register_all` would be sensible, while there's still value in shipping them with the gem (the gem could track endpoint and URL pattern changes for everybody). So something like

```rb
OEmbed::Providers.register OEmbed::Providers::Instagram.new(access_token: "#{foo}")
```

**Comment by chris-hinds** [comment]
Thanks for the swift response @metavida I can totally see why you wouldn't include your own access token for use in this library, that would just be bad practice anyway let alone open up a whole heap of security issues.

Is the ability to register our own providers already present? so in theory I can register my own Instagram provider with my own access token and use this today?

I think however it would also be nice to simply be able to provide some env variables as you say and perhaps the library can detect if those env variables are set. If they are not set perhaps it errors stating you need the access token or you need to create the provider yourself?

**Comment by Maxhou00** [comment]
@metavida Anyone to accept the pullrequest #75 ? the implementation is pretty good and I think it's a major update

**Comment by metavida** [comment]
Thanks for the nudge! The PR arrived at a bad/busy time and had since
fallen off my radar. I just now looked over the diff on my phone and think
it’ll be pretty quick/easy to merge, especially since the API is backwards
compatible! A few small things to get in place (like documentation) before
it’d be release ready. I can’t guarantee a release date but will try to
make time for it soon.

**Comment by metavida** [comment]
I've released [ruby-oembed 0.14.0](https://rubygems.org/gems/ruby-oembed/versions/0.14.0) which offers two options to specify your [Facebook access token](https://developers.facebook.com/docs/instagram/oembed#access-tokens) (therefore fixing Instagram & Facebook oEmbed providers).

Option 1: Pass `access_tokens` to `register_all`

```ruby
OEmbed::Providers.register_all(access_tokens: { facebook: @my_facebook_token })
```

Option 2: Use environment variables

```ruby
ENV['OEMBED_FACEBOOK_TOKEN'] #=> 'my-access-token'
OEmbed::Providers.register_all
```

A huge thanks to @elektronaut for the PR! And thanks all for your continued suggestions & use of this gem ❤️
